src/tests/end2end/DepthStencilStateTests.cpp

```diff
@@ -0,0 +1,512 @@
+        }
+
+        void DoTest(const std::vector<TestSpec> &testParams, const RGBA8& expected) {
+            DoTest(testParams, expected, expected);
+        }
+
+        nxt::RenderPass renderpass;
+        nxt::Texture renderTarget;
+        nxt::Texture depthTexture;
+        nxt::TextureView renderTargetView;
+        nxt::TextureView depthTextureView;
+        nxt::Framebuffer framebuffer;
+        nxt::ShaderModule vsModule;
+        nxt::ShaderModule fsModule;
+        nxt::BindGroupLayout bindGroupLayout;
+        nxt::PipelineLayout pipelineLayout;
+};
+
+// Test compilation and usage of the fixture
+TEST_P(DepthStencilStateTest, Basic) {
+    nxt::DepthStencilState state = device.CreateDepthStencilStateBuilder()
+        .GetResult();
+
+    DoTest({
+        { state, RGBA8(0, 255, 0, 255), 0.5f, 0u },
+    }, RGBA8(0, 255, 0, 255));
+}
+
+// Test defaults: depth and stencil tests disabled
+TEST_P(DepthStencilStateTest, Default) {
+    nxt::DepthStencilState state = device.CreateDepthStencilStateBuilder()
+        .GetResult();
+    
+    TestSpec specs[3] = {
+        { state, RGBA8(255, 0, 0, 255), 0.0f, 0u },
+        { state, RGBA8(0, 255, 0, 255), 0.5f, 0u },
+        { state, RGBA8(0, 0, 255, 255), 1.0f, 0u },
+    };
+
+    // Test that for all combinations, the last triangle drawn is the one visible
+    // We check against three triangles because the stencil test may modify results
+    for (uint32_t last = 0; last < 3; ++last) {
+        uint32_t i = last != 0 ? 0 : 1;
+        uint32_t j = 3 - (i + last);
+        DoTest({ specs[i], specs[j], specs[last] }, specs[last].color);
+        DoTest({ specs[j], specs[i], specs[last] }, specs[last].color);
+    }
+}
+
+TEST_P(DepthStencilStateTest, DepthAlways) {
+    CheckDepthCompareFunction(nxt::CompareFunction::Always , true, true, true);
+}
+
+TEST_P(DepthStencilStateTest, DepthEqual) {
+    CheckDepthCompareFunction(nxt::CompareFunction::Equal, false, true, false);
+}
+
+TEST_P(DepthStencilStateTest, DepthGreater) {
+    CheckDepthCompareFunction(nxt::CompareFunction::Greater, false, false, true);
+}
+
+TEST_P(DepthStencilStateTest, DepthGreaterEqual) {
+    CheckDepthCompareFunction(nxt::CompareFunction::GreaterEqual, false, true, true);
+}
+
+TEST_P(DepthStencilStateTest, DepthLess) {
+    CheckDepthCompareFunction(nxt::CompareFunction::Less, true, false, false);
+}
+
+TEST_P(DepthStencilStateTest, DepthLessEqual) {
+    CheckDepthCompareFunction(nxt::CompareFunction::LessEqual, true, true, false);
+}
+
+TEST_P(DepthStencilStateTest, DepthNever) {
+    CheckDepthCompareFunction(nxt::CompareFunction::Never, false, false, false);
+}
+
+TEST_P(DepthStencilStateTest, DepthNotEqual) {
+    CheckDepthCompareFunction(nxt::CompareFunction::NotEqual, true, false, true);
+}
+
+// Draw a base triangle with depth enabled, draw a second triangle without depth enabled, draw a third triangle which chould occlude the second
```

Comment:
nit: this comment can go in the test, before the test we should say what the test checks for.